### DISCUSSION TITLE
Question on generated cpp for constexpr object

### DISCUSSION THREAD

**Discussion by mike919192** [hsutter/cppfront#731] [Q&A] [answered]
Hi.  I finally got a chance today to try out the constexpr functionality that was added recently.  Functions worked exactly as I expected, but I was surprised by the generated cpp for constexpr objects.

This is the example from the last update

```
// object with "always the same value" (constexpr object)
forty_two: i64 == 42;
```

compiles to this cpp code

```
//actual cpp code
cpp2::i64 const& forty_two = 42;
```

but I was expecting something more like this cpp code

```
//expected cpp code
constexpr cpp2::i64 forty_two{42};
```

The actual code cannot be used where constexpr is required.  For example:

```
//produces error
test_array: std::array<uint, forty_two> = 0;
```

```
hello.cpp2: In function ‘int main()’:
hello.cpp2:11:21: error: the value of ‘forty_two’ is not usable in a constant expression
   11 |     test: std::array<uint, forty_two> = 0;
      |                     ^~~~~~~~~
hello.cpp2:10:22: note: ‘forty_two’ was not initialized with a constant expression
   10 |     forty_two: i64 == 42;
      |                      ^~~~     
hello.cpp2:11:30: error: the value of ‘forty_two’ is not usable in a constant expression
   11 |     test: std::array<uint, forty_two> = 0;
      |                              ^
hello.cpp2:10:22: note: ‘forty_two’ was not initialized with a constant expression
   10 |     forty_two: i64 == 42;
      |                      ^~~~     
hello.cpp2:11:21: note: in template argument for type ‘long unsigned int’
   11 |     test: std::array<uint, forty_two> = 0;
```

Is this the intended design?  Because it doesn't seem to be usable when constexpr is required.

**Comment by JohelEGP** [comment, marked as answer]
This is the intended design, although perhaps outdated.
At <https://github.com/hsutter/cppfront/commit/b589f5d25e5acdbfd94791e23c0ba0f6fdcd59c6#commitcomment-129028450>, I argue that this should also lower to a `constexpr` object.

**Reply by mike919192** [reply]
Ok thanks for the reply.  So is it correct to say that there is currently no pure cpp2 way to write this?

`constexpr cpp2::i64 forty_two{42};`

Seems like the best that is currently possible would be to used mixed cpp / cpp2.  But that is also limited because I guess you would not be able to have a constexpr object declared inside of a cpp2 function.

**Reply by JohelEGP** [reply]
You can do this: <https://github.com/hsutter/cppfront/issues/568#issuecomment-1662251468>.

**Reply by JohelEGP** [reply]
> So is it correct to say that there is currently no pure cpp2 way to write this?

That's right.

This is an alternative (<https://cpp2.godbolt.org/z/GsM5hj7zd>):
```Cpp2
constant: <V: _> type == std::integral_constant<decltype(V), V>;
main: () = {
  ROWS: type == constant<3>;
  static_assert(ROWS() == 3);
  static_assert(ROWS::value == 3);
}
```

**Reply by mike919192** [reply]
Thanks.  That's usable as a workaround.  I was just very confused when I first tried the code from the Autumn update since it says its supposed to be a constexpr object but isnt really, at least for now.  Maybe the implementation is still WIP.

**Comment by hsutter** [comment]
> forty_two: i64 == 42;

Thanks! This is just an oversight. Will fix.

Note that when the above is a non-local variable, it does lower to a `constexpr` variable: `cpp2::i64 inline constexpr forty_two = 42;`

When I made the change, I meant (but forgot) to check that the same (except not `inline`) was now being done for local variables for consistency.

Note that if you do want the equivalent of a local `const&`, you can still write it using block/statement scope variables which default to `in` semantics:

    (v2 := v)  // v2 is a vector<T> const&
    for v2 do (s)
        std::cout << "(s)$\n";

**Reply by hsutter** [reply]
Fixed in: 3512ecd41f33f7e812a070e3370fd97845dd7d24

Thanks again.
